base/linalg/rowvector.jl

```diff
@@ -0,0 +1,183 @@
+        throw(BoundsError(rowvec, i.I))
+    end
+    @inbounds rowvec.vec[i.I[2]] = transpose(v)
+end
+
+@propagate_inbounds getindex(rowvec::RowVector, ::CartesianIndex{0}) = getindex(rowvec)
+@propagate_inbounds getindex(rowvec::RowVector, i::CartesianIndex{1}) = getindex(rowvec, i.I[1])
+
+@propagate_inbounds setindex!(rowvec::RowVector, v, ::CartesianIndex{0}) = setindex!(rowvec, v)
+@propagate_inbounds setindex!(rowvec::RowVector, v, i::CartesianIndex{1}) = setindex!(rowvec, v, i.I[1])
+
+@inline check_tail_indices(i1, i2) = true
+@inline check_tail_indices(i1, i2, i3, is...) = i3 == 1 ? check_tail_indices(i1, i2, is...) :  false
+
+# Some conversions
+convert(::Type{AbstractVector}, rowvec::RowVector) = rowvec.vec
+convert{V<:AbstractVector}(::Type{V}, rowvec::RowVector) = convert(V, rowvec.vec)
+convert{T,V}(::Type{RowVector{T,V}}, rowvec::RowVector) = RowVector(convert(V, rowvec.vec))
+
+# helper function for below
+@inline to_vec(rowvec::RowVector) = transpose(rowvec)
+@inline to_vec(x::Number) = x
+@inline to_vecs(rowvecs...) = (map(to_vec, rowvecs)...)
+
+# map
+@inline map(f, rowvecs::RowVector...) = RowVector(map(f, to_vecs(rowvecs...)...))
+
+# broacast (other combinations default to higher-dimensional array)
+@inline broadcast(f, rowvecs::Union{Number,RowVector}...) = RowVector(broadcast(f, to_vecs(rowvecs...)...))
+
+# Horizontal concatenation #
+
+# Empty hcat defaults to row vector
+hcat() = transpose([])
+@inline hcat{T}(X::T...)         = transpose(transpose_type(T)[ transpose(X[j]) for j=1:length(X) ])
+@inline hcat{T<:Number}(X::T...) = transpose(T[ X[j] for j=1:length(X) ])
+@inline hcat(X::Number...) = transpose(hvcat_fill(Array{promote_typeof(X...),1}(length(X)), X))
+@inline hcat(X::RowVector...) = transpose(vcat(map(transpose, X)...))
+@inline hcat(X::Union{RowVector,Number}...) = transpose(vcat(map(transpose, X)...))
+
+typed_hcat{T}(::Type{T}) = transpose(typed_vcat(T))
+@inline typed_hcat{T}(::Type{T}, X::Number...) = transpose(hvcat_fill(Array{T,1}(length(X)), X))
+@inline typed_hcat{T}(::Type{T}, X::RowVector...) = transpose(typed_vcat(T, map(transpose, X)...))
+@inline typed_hcat{T}(::Type{T}, X::Union{RowVector,Number}...) = transpose(typed_vcat(T, map(transpose, X)...))
+
+# Multiplication #
+
+@inline *(rowvec::RowVector, vec::AbstractVector) = reduce(+, map(At_mul_B, transpose(rowvec), vec))
+@inline *(rowvec::RowVector, mat::AbstractMatrix) = transpose(mat.' * transpose(rowvec))
+@inline *(vec::AbstractVector, mat::AbstractMatrix) = error("Cannot left-multiply a matrix by a vector") # Should become a deprecation
+@inline *(::RowVector, ::RowVector) = error("Cannot multiply two transposed vectors")
+@inline *(vec::AbstractVector, rowvec::RowVector) = kron(vec, rowvec)
+@inline *(vec::AbstractVector, rowvec::AbstractVector) = error("Cannot multiply two vectors")
+@inline *(mat::AbstractMatrix, rowvec::RowVector) = error("Cannot right-multiply matrix by transposed vector")
```

Comment:
Why are all of these error definitions necessary?  We already throw errors for multiplying matrices and vectors of incompatible shapes, and since `RowVector <: AbstractMatrix` the existing code should just work (i.e. continue to throw errors), no?